Generate tiny compiled binary for wrapping executables

bergkvist opened the pull request:
# Generate tiny compiled binary for wrapping executables

###### Try out the latest version

```sh
env NIX_PATH=nixpkgs=https://github.com/bergkvist/nixpkgs/archive/bergkvist/make-c-wrapper.tar.gz \
nix-shell -p makeBinaryWrapper --run "makeWrapper [EXECUTABLE] [OUT_PATH] [ARGS]"
```

###### Supported ARGS:

- `--argv0 NAME`: set name of executed process to NAME (defaults to EXECUTABLE)
- `--inherit-argv0`: the executable inherits argv0 from the wrapper. (use instead of `--argv0 '$0'`)
- `--set VAR VAL`: add VAR with value VAL to the executable’s environment
- `--set-default VAR VAL`: like --set, but only adds VAR if not already set in the environment
- `--unset VAR`: remove VAR from the environment
- `--chdir DIR`: change working directory (use instead of `--run "cd DIR"`)
- `--add-flags FLAGS`: add FLAGS to invocation of executable
- `--prefix ENV SEP VAL`: prefix ENV with VAL, separated by SEP
- `--suffix ENV SEP VAL`: suffix ENV with VAL, separated by SEP

###### Explore generated code
If you are curious about the generated code, and not interested in actually creating a wrapper, you can use the following, which will simply print out the C-code:
```sh
env NIX_PATH=nixpkgs=https://github.com/bergkvist/nixpkgs/archive/bergkvist/make-c-wrapper.tar.gz \
nix-shell -p makeBinaryWrapper --run "makeCWrapper [EXECUTABLE] [ARGS]"
```

###### Motivation for this change
Nix uses [bash-wrappers](https://github.com/NixOS/nixpkgs/blob/1d6428140194b5bac68266ca11a441ed6f63571c/pkgs/build-support/setup-hooks/make-wrapper.sh) to inject custom environment variables into executables. In the cases where this executable is an interpreter (like Python or Perl), it can be placed in the shebang line of a script. On MacOS, you can't put a script (with its own shebang) in the shebang line of an executable [due to a limitation with the `execve`-syscall](https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working). 

If we had some way of generating tiny compiled binaries instead of bash-scripts as wrappers then this would work on both Linux and MacOS. 

#### See https://github.com/NixOS/nixpkgs/issues/23018 for more discussion.

A focus in this PR has been to minimize the number of dependencies - and also keep the implementation itself as minimal as possible.

###### Dependencies
- `bash`
- `awk`
- A C-compiler: `gcc` on Linux or `clang` on MacOS
- [`unistd`](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/unistd.h.html) + [`stdlib`](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdlib.h.html) + [`assert`](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/assert.h.html) + [`stdio`](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdio.h.html) (C POSIX libraries)
- [`asprintf`](https://man7.org/linux/man-pages/man3/asprintf.3.html) (GNU extension for stdio which is available on macOS/OpenBSD 2.3+)

## Consider the following wrapper shell script:
```sh
#! /nix/store/ra8yvijdfjcs5f66b99gdjn86gparrbz-bash-4.4-p23/bin/bash -e
export NIX_PYTHONPREFIX='/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env'
export NIX_PYTHONEXECUTABLE='/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env/bin/python3.9'
export NIX_PYTHONPATH='/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env/lib/python3.9/site-packages'
export PYTHONNOUSERSITE='true'
exec "/nix/store/7pjbbmnrch7frgyp7gz19ay0z1173c7y-python3-3.9.2/bin/python"  "$@"
```
Putting this script in a shebang works fine on Linux, but doesn't work on MacOS. If we want to write C-code that replaces it (and works on MacOS+Linux), we can do it like this in C:
```c
#include <unistd.h>
#include <stdlib.h>

int main(int argc, char** argv) {
    putenv("NIX_PYTHONPREFIX=/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env");
    putenv("NIX_PYTHONEXECUTABLE=/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env/bin/python3.9");
    putenv("NIX_PYTHONPATH=/nix/store/i46k148mi830riq4wxh49ki8qmq0731k-python3-3.9.2-env/lib/python3.9/site-packages");
    putenv("PYTHONNOUSERSITE=true");
    argv[0] = "/nix/store/7pjbbmnrch7frgyp7gz19ay0z1173c7y-python3-3.9.2/bin/python";
    return execv(argv[0], argv);
}
```

This PR has creates a simple bash function that generates C-code for such a tiny compiled binary (and compiles it), with an interface similar to the existing makeWrapper. There are some features of the original makeWrapper which is not yet implemented here.

### Debuggability

A big concern with using a binary wrapper is that people can't just open up the file to see what it is doing when debugging their own problems. This is fixed by embedding the arguments used to create the wrapper as a string variable into the source code itself. The result is that the binary file will contain the wrapper arguments in human readable format when opening the file in a plain text editor or using the `strings` command on MacOS or Linux.

Example of how it looks right now:
```sh
makeBinaryWrapper /usr/bin/python3 ./wrapper \
  --set HELLO WORLD --set-default X $'Y\n"' --unset Z --argv0 python3

cat ./wrapper
```

```sh
...binary-data...
# ------------------------------------------------------------------------------------
# The C-code for this binary wrapper has been generated using the following command:


makeCWrapper /usr/bin/python3 \
    --set 'HELLO' 'WORLD' \
    --set-default 'X' $'Y\n"' \
    --unset 'Z' \
    --argv0 'python3'


# (Use `nix-shell -p makeBinaryWrapper` to get access to makeCWrapper in your shell)
# ------------------------------------------------------------------------------------
...binary-data...
```

C String literals in the generated code (including the documentation) are properly escaped. I got some help with how to do this properly on StackOverflow: https://stackoverflow.com/questions/67710149/how-can-i-sanitize-user-input-into-valid-c-string-literals.

### Generated source code example
```sh
makeDocumentedCWrapper /usr/bin/python3 --set HELLO WORLD --set-default X $'Y\n"' --unset Z --argv0 python3
```
```c
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>

#define assert_success(e) do { if ((e) < 0) { perror(#e); abort(); } } while (0)

int main(int argc, char **argv) {
    putenv("HELLO=WORLD");
    assert_success(setenv("X", "Y\n\"", 0));
    assert_success(unsetenv("Z"));
    argv[0] = "python3";
    return execv("/usr/bin/python3", argv);
}

const char * DOCSTRING = "\n\n\n# ------------------------------------------------------------------------------------\n# The C-code for this binary wrapper has been generated using the following command:\n\n\nmakeCWrapper /usr/bin/python3 \\\n    --set 'HELLO' 'WORLD' \\\n    --set-default 'X' $'Y\\n\"' \\\n    --unset 'Z' \\\n    --argv0 'python3'\n\n\n# (Use `nix-shell -p makeBinaryWrapper` to get access to makeCWrapper in your shell)\n# ------------------------------------------------------------------------------------\n\n\n";
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

## Issues that this PR would likely impact:
- https://github.com/NixOS/nixpkgs/issues/2146
- https://github.com/NixOS/nixpkgs/issues/11133
- https://github.com/NixOS/nixpkgs/issues/23018
- https://github.com/NixOS/nixpkgs/issues/65351
- https://github.com/NixOS/nixpkgs/issues/86881
- https://github.com/NixOS/nixpkgs/issues/93609
- https://github.com/NixOS/nixpkgs/issues/123067

bergkvist committed:
Add bergkvist to maintainer list

bergkvist committed:
Add make-c-wrapper.sh for creating binary executable wrappers

bergkvist committed:
make-c-wrapper: Fix typo in generated code documentation

bergkvist committed:
make-c-wrapper: Remove trailing whitespace (editorconfig)

bergkvist committed:
make-c-wrapper: Fix typo in generated documentation

xaverdh commented:
see https://github.com/NixOS/nixpkgs/pull/95569 for prior art

bergkvist committed:
Rename make-c-wrapper.sh to make-binary-wrapper.sh. Refactor to match style of other setup-hooks. Add compilation step with gcc. Embed the entire generated source code into the binary for troubleshooting.

bergkvist committed:
Remove line at the bottom of make-binary-wrapper that executes makeBinaryWrapper "$@"

hrhino commented:
I love this! It took a moment of fiddling about but I used the default `makeWrapper`, this PR's `makeBinaryWrapper`, and `makeBinaryWrapper` from #95569, and got this:
```
-r-xr-xr-x 1 root root  80128 Dec 31  1969 binary-nim
-r-xr-xr-x 1 root root    291 Dec 31  1969 current
-rwxr-xr-x 1 haro users 14232 Dec 31  1969 binary-c
```
so while this is significantly heavier than the shell-script version (unsurprisingly &mdash; the debug string literal is going to be around the size of the script) it's _significantly_ smaller than the one that depends on `nim` and its json libraries and such, so I think it's seriously worth considering this over that (the operation of this executable is also more straightforward IMO).

On the other hand, that one has much more features. Looking at the current uses of `makeWrapper`, I think you'll have to add `--add-flags`, `--prefix`, and `--suffix` at least to support most use cases.

bergkvist committed:
Add support for --add-flags, --prefix and --suffix

bergkvist commented:
@hrhino Thanks for the feedback!
I've added --add-flags, --prefix and --suffix now, without impacting the resulting binary size too much. They were slightly more involved than the previous ones - but hopefully the generated code should still be relatively easy enough to read/understand.

I turned on -Os (optimize for small size) to keep the binary size small. It didn't help much in making the binaries smaller - but it helped in preventing them from getting bigger when I added the functions I use for --prefix and --suffix.

Although the compiler will automatically discard functions that are not being used - the debug string won't - so in order to keep the debug string clean - I only include the functions used by --prefix and --suffix if they are actually used.

bergkvist committed:
Switch to using strlen in concat3Fn. Make sure uses-variables are local to improve purity of makeCWrapper. Refactor

roberth reviewed (commented):
This looks promising.

Could you add tests?

 - some examples to test the behavior
 - some means of ensuring memory safety
    - running the examples in `valgrind`; might not catch everything
    - running a static analysis on the generated code for the examples (don't know which tool to use. Any suggestions?)
    - adding some golden tests, comparing the generated code to known-correct generated code. This way memory safety of the generated code can be checked by hand, at least for some inputs
 
You can add a derivation that runs these tests in the same directory as the implementation and `callPackage` it from `pkgs/test/default.nix`.

SuperSandro2000 commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+}
+
+concat3Fn() {
+    printf "%s\n" 'char *concat3(char *x, char *y, char *z) {'
+    printf "%s\n" '    int xn = strlen(x);'
+    printf "%s\n" '    int yn = strlen(y);'
+    printf "%s\n" '    int zn = strlen(z);'
+    printf "%s\n" '    char *res = malloc(sizeof(*res)*(xn + yn + zn + 1));'
+    printf "%s\n" '    for (int i = 0; i < xn; ++i) res[i] = x[i];'
+    printf "%s\n" '    for (int i = 0; i < yn; ++i) res[xn+i] = y[i];'
+    printf "%s\n" '    for (int i = 0; i < zn; ++i) res[xn+yn+i] = z[i];'
+    printf "%s\n" "    res[xn+yn+zn] = '\0';"
+    printf "%s\n" '    return res;'
+    printf "%s\n" '}'
```
Why not use multiline strings or EOF instead?

bergkvist replied:
The reason I've been using printf over echo is that it is supposedly more standardized/more portable (https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo/65819#65819)

That said, the edge cases where this is true might not apply/make a difference here.

I've switched to using a multiline string instead of multiple printf-lines as you suggested

SuperSandro2000 commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+
+# setDefaultEnv KEY VALUE
+setDefaultEnv() {
+    local key=$(escapeStringLiteral "$1")
+    local value=$(escapeStringLiteral "$2")
```
shellcheck will be pretty unhappy about all of those. Please check the script with it.

ncfavier replied:
There's no issue here. Look for "assignment statements" in https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameters

Shellcheck does warn about masking return values, but it's not like `escapeStringLiteral` can fail.

SuperSandro2000 replied:
> There's no quoting issue here.

I never talked about quoting. I was referring to return value masking. Use shellcheck to get the exact details.

bergkvist replied:
Yeah, I get this warning from shellcheck:
```
Declare and assign separately to avoid masking return values.shellcheck(SC2155)
```

If I understand correctly, I need to change from

```bash
local key=$(escapeStringLiteral "$1")
local value=$(escapeStringLiteral "$2")
```

to:

```bash
local key value
key=$(escapeStringLiteral "$1")
value=$(escapeStringLiteral "$2")
```

SuperSandro2000 reviewed (commented):
Please consider using echo instead of printf with a new line at the end.

ncfavier commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+makeDocumentedCWrapper() {
+    local src=$(makeCWrapper "$@")
+    local docs=$(documentationString "$src")
+    printf "%s\n" "$src"
```
This applies to the whole file: prefer `"\\n"` or `'\n'` (it could be argued that `"\n"` is an invalid escape sequence, but Bash just ignores it). For printf format strings, single quotes are usually nicer.

bergkvist replied:
So if I understand correctly, I should change
```bash
printf "%s\n" "$src"
```

to
```bash
printf '%s\n' "$src"
```

across the entire file?

ncfavier replied:
Yes

ncfavier commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+
+    for ((n = 1; n < ${#params[*]}; n += 1)); do
+        p="${params[$n]}"
+        if [[ "$p" == "--set" ]]; then
```
Use a `case` statement:
```bash
case $p in # no quoting needed here
    --set) …;;
    …
    *) …;;
esac
```

ncfavier commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+    local executable=$(escapeStringLiteral "$1")
+    local params=("$@")
+
+    for ((n = 1; n < ${#params[*]}; n += 1)); do
```
Why are you skipping `n = 0`?

ncfavier commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+    local params=("$@")
+
+    for ((n = 1; n < ${#params[*]}; n += 1)); do
+        p="${params[$n]}"
```
`${params[n]}`

ncfavier commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,198 @@
+    for ((n = 1; n < ${#params[*]}; n += 1)); do
+        p="${params[$n]}"
+        if [[ "$p" == "--set" ]]; then
+            cmd=$(setEnv "${params[$((n + 1))]}" "${params[$((n + 2))]}")
```
`${params[n + 1]}`, `${params[n + 2]}`

bergkvist commented:
### In terms of memory safety

As long as the wrapper is executed as its own process, the OS will take care of freeing any remaining memory when the process exits. So if there are memory leaks in the wrapper, it will simply add an additional constant overhead to the process that has been wrapped.

It is not possible to free things like the `argv` that was passed to main, even if the child process does not use it - so some memory overhead from wrapping seems unavoidable.

We can only safely free memory that is not used by the wrapped program. That essentially means only "temporary" variables can be freed. From the code I've written so far, the only temporary variable that is not consumed by the wrapped process is the result of `char *concat3(char *x, char *y, char *z);`.

The reason `val` can be freed after calling `setenv(key, val, 1);` is that setenv creates its own internal copy of `val`. Since we are not allowed to free environment variables passed from the parent process, we must make sure to only free `val` if it is the result of `char *concat3(char *x, char *y, char *z);`

roberth commented:
I'm not concerned about _memory leaks_ either, but I am concerned about _undefined behavior_.

Seemingly innocent mistakes can lead to security problems, amplified by the fact that such a simple core component would be assumed to be trustworthy by many.

Even just the "golden test" idea goes a long way when it comes to this responsibility, because it will let maintainers check the generated programs manually and help us check them again when the code generation inevitably needs to change.

bergkvist committed:
Fix shellcheck warnings. Use single quotes for printf format strings. Switch to strncpy in concat3. Use multiline strings to print C functions. Switch from if/elif to case.

doronbehar commented:
This looks so great :) A few comments, first of all:

```diff
diff --git i/pkgs/top-level/all-packages.nix w/pkgs/top-level/all-packages.nix
index c53485f5dea..30b38909952 100644
--- i/pkgs/top-level/all-packages.nix
+++ w/pkgs/top-level/all-packages.nix
@@ -642,6 +642,8 @@ with pkgs;
   makeWrapper = makeSetupHook { deps = [ dieHook ]; substitutions = { shell = targetPackages.runtimeShell; }; }
                               ../build-support/setup-hooks/make-wrapper.sh;
 
+  makeBinaryWrapper = makeSetupHook { } ../build-support/setup-hooks/make-binary-wrapper.sh;
+
   makeModulesClosure = { kernel, firmware, rootModules, allowMissing ? false }:
     callPackage ../build-support/kernel/modules-closure.nix {
       inherit kernel firmware rootModules allowMissing;
```

And then it's easy to test this with:

```
env NIX_PATH=nixpkgs=$PWD nix-shell -p makeBinaryWrapper --run makeBinaryWrapper <args>
```

Or even without the need to clone and checkout your branch:

```
env NIX_PATH=nixpkgs=https://github.com/bergkvist/nixpkgs/archive/bergkvist/make-c-wrapper.tar.gz nix-shell -p makeBinaryWrapper --run makeBinaryWrapper <args>
```

It'd be nice to put that info in the first comment of the thread :).

Other then that, When giving a try to `--prefix`, I forgot to give 3 arguments (forgot the separator) and it took me a while to realize it failed silently to wrap properly the executable - it'd be nice if the bash function would have checked it received a proper number of arguments for each option.

bergkvist committed:
Add makeBinaryWrapper to pkgs/top-level/all-packages.nix

bergkvist committed:
Return an #error macro if the wrong number of arguments are supplied

bergkvist commented:
@doronbehar Thanks for the suggestion! I added it to pkgs/top-level/all-packages.nix, and put the command you suggested at the top of the first comment of this PR.

Detecting whether the wrong number of arguments are passed could be tricky - especially if it is not the last element. That said, I added a check to make it not fail silently if the number of arguments don't add up.

bergkvist committed:
Make error messages more consistent

doronbehar committed:
makeBinaryWrapper: Assert $1 is executable

doronbehar committed:
makeBinaryWrapper: add wrapProgramBinary (like wrapProgram)

doronbehar committed:
makeBinaryWrapper: Document

bergkvist committed:
Specify uses_prefix, uses_suffix and uses_concat3 as local vars. Make sure errors in makeCWrapper are located next to the line with the issue.

bergkvist committed:
Add golden tests for make-binary-wrapper.

To run tests after cloning on linux, use the following:
nix-build pkgs/top-level/release.nix -A tests.make-binary-wrapper.x86_64-linux

bergkvist committed:
EditorConfig: Switch from tabs to spaces

bergkvist commented:
## Golden tests
I've created a system for writing golden tests for makeBinaryWrapper. I added tests to at least cover every type of argument. For the time being, I copy-pasted some of the testing logic from the patch-shebangs-test.

### Running the tests on Linux
```sh
git clone https://github.com/bergkvist/nixpkgs bergkvist-nixpkgs
cd bergkvist-nixpkgs
git checkout bergkvist/make-c-wrapper
nix-build pkgs/top-level/release.nix -A tests.make-binary-wrapper.x86_64-linux
```

### Example test
```c
// makeCWrapper /hello/world \
    --set PART1 HELLO \
    --set-default PART2 WORLD \
    --unset SOME_OTHER_VARIABLE \
    --set PART3 $'"!!\n"'

#include <unistd.h>
#include <stdlib.h>

int main(int argc, char **argv) {
    putenv("PART1=HELLO");
    setenv("PART2", "WORLD", 0);
    unsetenv("SOME_OTHER_VARIABLE");
    putenv("PART3=\"!!\n\"");
    argv[0] = "/hello/world";
    return execv("/hello/world", argv);
}
```

Notice that the first comment serves as the input command (after `//` has been stripped away). Having both the input command and the generated code in the same file makes it easier to create new tests or modify existing ones.

The backslashes serve a dual purpose:
- Allow for writing the input command using more than one line
- Ensure that the input command remains a comment when using C syntax highlighting, in spite of being spread across multiple lines.

bergkvist committed:
Move assertExecutable from makeCWrapper to makeBinaryWrapper to ensure that makeCWrapper is a pure function

doronbehar committed:
Improve explenations for wrap*Program

bergkvist committed:
Set strictDeps = true in makeGoldenTest

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

bergkvist committed:
Switch from buildInputs to nativeBuildInpuits in makeGoldenTest

doronbehar committed:
Rephrase makeWrapper setup-hook

bergkvist commented:
In terms of memory safety - as an alternative to [CodeHawk-C](https://github.com/static-analysis-engineering/CodeHawk-C), we could also consider using [`infer` by Facebook](https://fbinfer.com/). *(Downside: not as lightweight)*

Here is an example to demonstrate null pointer dereferencing and buffer overrun issues:
```c
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
    char **argv_tmp = malloc(sizeof(*argv_tmp) * (4 + argc));
    argv_tmp[0] = argv[0];
    argv_tmp[1] = "-x";
    argv_tmp[2] = "-y";
    argv_tmp[3] = "-z";
    for (int i = 1; i < argc; ++i) {
        argv_tmp[5 + i] = argv[i];
    }
    argv_tmp[4 + argc] = NULL;
    argv = argv_tmp;

    argv[0] = "my-wrapper";
    return execv("/path/to/executable", argv);
}
```

Command:
`infer run` [`--biabduction`](https://fbinfer.com/docs/checker-biabduction) [`--bufferoverrun`](https://fbinfer.com/docs/checker-bufferoverrun) `--` `gcc -c main.c`
```c
Capturing in make/cc mode...
Found 1 source file to analyze in /home/tobias/infer-test/infer-out
1/1 [################################################################################] 100% 167ms

main.c:14: error: Buffer Overrun L1
  Offset: `argc` + 4 Size: `argc` + 4.
  12.         argv_tmp[5 + i] = argv[i];
  13.     }
  14.     argv_tmp[4 + argc] = NULL;
          ^
  15.     argv = argv_tmp;
  16. 

main.c:12: error: Buffer Overrun L2
  Offset: [6, `argc` + 4] Size: `argc` + 4.
  10.     argv_tmp[3] = "-z";
  11.     for (int i = 1; i < argc; ++i) {
  12.         argv_tmp[5 + i] = argv[i];
              ^
  13.     }
  14.     argv_tmp[4 + argc] = NULL;

main.c:7: error: Null Dereference
  pointer `argv_tmp` last assigned on line 6 could be null and is dereferenced at line 7, column 5.
  5. int main(int argc, char **argv) {
  6.     char **argv_tmp = malloc(sizeof(*argv_tmp) * (4 + argc));
  7.     argv_tmp[0] = argv[0];
         ^
  8.     argv_tmp[1] = "-x";
  9.     argv_tmp[2] = "-y";


Found 3 issues
            Issue Type(ISSUED_TYPE_ID): #
    Null Dereference(NULL_DEREFERENCE): 1
  Buffer Overrun L2(BUFFER_OVERRUN_L2): 1
  Buffer Overrun L1(BUFFER_OVERRUN_L1): 1
```

In order to fix these issues, I can then do:
```c
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

int main(int argc, char **argv) {
    char **argv_tmp = malloc(sizeof(*argv_tmp) * (4 + argc));
    assert(argv_tmp != NULL);
    argv_tmp[0] = argv[0];
    argv_tmp[1] = "-x";
    argv_tmp[2] = "-y";
    argv_tmp[3] = "-z";
    for (int i = 1; i < argc; ++i) {
        argv_tmp[3 + i] = argv[i];
    }
    argv_tmp[3 + argc] = NULL;
    argv = argv_tmp;

    argv[0] = "my-wrapper";
    return execv("/path/to/executable", argv);
}
```

```c
Capturing in make/cc mode...
Found 1 source file to analyze in /home/tobias/infer-test/infer-out
1/1 [################################################################################] 100% 227ms

  No issues found  
```

### A relevant takeaway:

Infer static analysis wants us to check that the returned value from malloc is not a NULL pointer. If malloc fails, we probably want to crash anyways - so we can use assert to be explicit about this behaviour. If malloc actually fails - then we will also get much better error messages when using assert:
```
[1]    1020278 segmentation fault (core dumped)  ./a.out
```

vs

```
a.out: main.c:8: main: Assertion `argv_tmp != NULL' failed.
[1]    1020411 abort (core dumped)  ./a.out
```

bergkvist committed:
Assert that malloc does not return a NULL pointer for better error messages + to satisfy static analysis tools.

andrew-d commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,260 @@
+    printf '%s' "$result"
+}
+
+concat3Fn() {
```
Seems to work fine on my M1 Pro, FWIW:
```
$ uname -sm
Darwin arm64
$ cat test.c
#include <stdio.h>

int main() {
    char *x;
    int size = asprintf(&x, "%s%s%s", "12", "34", "56");

    printf("size = %d\n", size);
    printf("x    = %s\n", x);
}
$ gcc -o test test.c && ./test
size = 6
x    = 123456
$ file test
test: Mach-O 64-bit executable arm64
```

bergkvist replied:
I'm getting this warning on Manjaro Linux, although it seems to work:
```
test.c: In function 'main':
test.c:5:16: warning: implicit declaration of function 'asprintf'; did you mean 'vsprintf'? [-Wimplicit-function-declaration]
    5 |     int size = asprintf(&x, "%s%s%s", "12", "34", "56");
      |                ^~~~~~~~
      |                vsprintf
size = 6
x    = 123456
```

samueldr replied:
I assume it's because it needs:

```
       #define _GNU_SOURCE         /* See feature_test_macros(7) */
```

 - https://man7.org/linux/man-pages/man3/asprintf.3.html

bergkvist commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,260 @@
+    local key value
+    key=$(escapeStringLiteral "$1")
+    value=$(escapeStringLiteral "$2")
+    printf '%s' "setenv(\"$key\", \"$value\", 0);"
```
I'm guessing we want to crash the program if setenv fails, so how about something like this?
```sh
printf '%s' "assert(setenv(\"$key\", \"$value\", 0) == 0);"
```

As far as I can tell there are 2 reasons setenv might fail: insufficient memory and variable name containing `=`, which we can catch during code generation instead of during runtime.

pennae replied:
that sounds reasonable. though if we're adding enough asserts it might also be nice to have a version of it that also prints errno, basically `#define assert_success(e) do { if ((e) < 0) { perror(#e); exit(1); } while (0)`

samueldr replied:
I `exit(1)` *equivalent enough* with an `assert()` return value?

I'm reading it's likely to be 134 (128 + SIGABRT), tbf I haven't tested.

Is there value in ensuring it exits in a similar manner?

pennae replied:
it's more likely to be equivalent to `raise(SIGABRT)`, which would indeed be better than using exit!

bergkvist committed:
Add assertValidEnvName and check that variable name is valid during code generation. Add assert_success, and assert that setenv/unsetenv succeeds to crash if they don't

pennae commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -133,6 +137,8 @@ makeCWrapper() {
     [ -z "$uses_concat3" ] || printf '%s\n' "#include <string.h>"
     [ -z "$uses_assert" ]  || printf '%s\n' "#include <assert.h>"
+    [ -z "$uses_assert_success" ] || printf '%s\n' "#include <stdio.h>"
+    [ -z "$uses_assert_success" ] || printf '\n%s\n' "#define assert_success(e) do { if ((e) < 0) { perror(#e); exit(1); } } while (0)"
```
as @samueldr noted, `abort()` would be better here than `exit(1)`

bergkvist committed:
Switch from exit(1) to abort() in assert_success

bergkvist committed:
Switch from malloc to calloc in addFlags

bergkvist committed:
Replace concat3 with asprintf in set_env_prefix and set_env_suffix.

bergkvist commented:
Thanks for the review @pennae! And for the helpful comments @samueldr and @andrew-d. I think I've addressed all of the latest comments now, so feel free to take another look if you want.

pennae commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,274 @@
+    mv "$prog" "$hidden"
+    # Silence warning about unexpanded $0:
+    # shellcheck disable=SC2016
+    makeBinaryWrapper "$hidden" "$prog" --argv0 '$0' "${@:2}"
```
```suggestion
    makeBinaryWrapper "$hidden" "$prog" --argv0 "$0" "${@:2}"
```
otherwise we get erroneous behaviour like this:
```
execve("/nix/store/sz6zhvzifmh8mj0frqmn0zmza2kzcirb-foo/bin/.foo-wrapped", ["$0"], 0x13712a0 /* 76 vars */) = 0
```

could you add a test for the generated executables as well?

pennae replied:
ah, now i see where this come from: seems it was lifted straight from `makeWrapper`, and thereby lost a level of expansion that would've happened in the generated shell script. the generated binary of course won't do this, causing this behaviour

bergkvist replied:
One option here could be to make `'$0'` a "magic value" when passed to --argv0 (making us not override argv[0]). That way, if existing scripts use this trick elsewhere with makeWrapper, they will continue to work with makeBinaryWrapper.

bergkvist replied:
The downside is that people might expect `'$1'` or `'$SOME_VAR'` to work as well, or try to pass `'$0'` to other things than --argv0

pennae replied:
the trick is that this `$0` ends up in the generated wrapper like so:
```
exec -a "$0" "/nix/store/rz6w8y7aci147cy3i2hiw5m7v9nyssm2-foo/bin/.foo-wrapped"  "$@"
```
would argue that all other uses of this trick (eg `--set FOO '$BAR'`) are highly suspect, and adding special casing to enable one specific instance seems like a detriment rather than an improvement. not modifying argv[0] would be fine, but it should be predicated on `--argv0` not being given instead of on the value it receives

bergkvist replied:
If you don't pass `--argv0` to the existing makeWrapper, it will be set to the name of the wrapped executable. Having makeBinaryWrapper behave differently from makeWrapper here might not be ideal. An alternative to allowing a magic value '$0' is to add another flag. Something like `--inherit-argv0` which takes 0 extra arguments.

pennae replied:
`--inherit-argv0` sounds good. :+1:

pennae reviewed (commented):
one more bug, otherwise just a few more nitpicks that may as well all be ignored. looking good!

bergkvist commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -790,6 +790,14 @@ There’s many more kinds of arguments, they are documented in `nixpkgs/pkgs/bui
 
+### `makeBinaryWrapper` \<executable\> \<wrapperfile\> \<args\> {#fun-makeBinaryWrapper}
+
+A setup-hook very similar to `makeWrapper`, only it creates a tiny _compiled_ wrapper executable, that can be used as a shebang interpreter. This is needed mostly on Darwin, where shebangs cannot point to scripts, [due to a limitation with the `execve`-syscall](https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working). The arguments it accepts are <!-- TODO: make  ...
```
Yeah, `system(...)` won't be able to replace `--run "cd som-dir"`, since using system is more analogous to wrapping the command in a subshell, while with the old `--run` things are executed at the "top level". Might make sense to add a flag [`--chdir DIR`](https://man7.org/linux/man-pages/man2/chdir.2.html) which replaces the usecase you describe.

bergkvist replied:
I'm a bit conflicted between `--set-workdir DIR` and `--chdir DIR` now. What do you think?

Some examples of `--run` in nixpkgs which would be hard to recreate in makeBinaryWrapper:

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/development/libraries/gstreamer/core/default.nix#L99 (changes environment inside --run)

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/tools/graphics/gnuplot/default.nix#L60
(sources a script to change the environment inside --run)

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/tools/text/popfile/default.nix#L45
(changes environment inside --run in a way which depends on other environment variables)

pennae replied:
> I'm a bit conflicted between --set-workdir DIR and --chdir DIR now. What do you think?

would go for `--chdir`

> Some examples of --run in nixpkgs which would be hard to recreate in makeBinaryWrapper:

not supporting those yet seems fine to me. if it turns out that it's necessary for something it'd be easy enough to add eg `--env-from-script <script>` that shall output a nul-separated list of `ENV=val` pairs the wrapper then sets in the environment. arguably all the missing actions of makeWrapper (except maybe the ability to chdir) fall in that same category of "not needed yet"

doronbehar replied:
According to what I see in `git grep -A3 '\(makeWrapper\|wrapProgram\)' | grep -- --run` the vast majority of usages for `--run` is for `cd` and for making the wrappers add paths for env vars relative to `$HOME`. How about adding a flag: `--set-relative` and `--set-relative-default`? Perhaps also we can add flags such as `--run-c` which will add c code to be embedded?

bergkvist committed:
Use cc instead of gcc in makeBinaryWrapper

bergkvist committed:
Add argument --inherit-argv0 to replace use case `--argv0 '$0'`. Fix wrapProgramBinary by using this new argument

bergkvist commented on pkgs/build-support/setup-hooks/make-binary-wrapper.sh:
```diff
@@ -0,0 +1,274 @@
+makeDocumentedCWrapper() {
+    local src docs
+    src=$(makeCWrapper "$@")
+    docs=$(documentationString "$src")
```
I've actually had this thought myself. Depends on whether people want to see what happens "under the hood" or if they prefer seeing the higher level description.

If someone is debugging the difference between makeWrapper and makeBinaryWrapper for their program - being able to see the C code would definitely be an advantage.

Switching to embedding inputs (arguments) instead of outputs (C-code) could certainly simplify the code generation quite a bit, by removing the need for uses_-flags. Making the code look tidy is less important if it is not embedded in the executable.

That said, if someone actually needs to debug makeCWrapper, it is an advantage if the code is as easy to read as possible. So in that sense, there might still be an advantage to keep the code minimal/readable, even if it doesn't the binary size.

pennae replied:
> That said, if someone actually needs to debug the output of makeCWrapper, it is an advantage if the code is as easy to read as possible.

as you've noted in your opening post, you can still get at the code by running `makeCWrapper`. perhaps the best way is in between, something like
```
# the C code for this wrapper binary can be recreated with
nix-shell -p makeBinaryWrapper --run '
  makeCWrapper <programName> \
    --argv0 <argv0> \
    ... etc ...
'
```
debugging the actual *wrapper* probably isn't a common operation, we *do* however inspect wrappers for what they do quite a lot.

bergkvist committed:
Add new argument: --chdir DIR (alternative to --run "cd DIR" in makeWrapper)

bergkvist committed:
Add tests for `--inherit-argv0` and `--chdir DIR`

doronbehar committed:
Remove TODO in documentation

tfc commented:
@bergkvist hi i just had a look at this after @roberth sent me this PR because of a completely different topic. I am really interested in getting this in, but i would change a few things.

- how about making the golden tests not compare actual source code but test the *effect* on the wrapped binary/script: this way we would not need to change *all the tests* on every single change. Also, we want to test if the *effect* that the wrapper has on the wrapped app is correct - users don't care about the C code as long as it's correct.
- what do you think about splitting the C-code generating script: have a script that only emits the generated code, and have another script compile it. this would make it easier to feed the output to static analysis, add sanitizers for tests builds etc. etc., i don't think that the script needlessly mixes multiple concerns right now.

tfc commented:
regardless of how the tests look in the end, i think we should have as a minimum requirement that we compile the output at least with `-Wall -Wpedantic -Werror` and add `-fsanitize=address -fsanitize=undefined` and run the wrappers to see if any sanitizer triggers.

the sanitizer can be instructed to ignore memory leaks using

```
export ASAN_OPTIONS=detect_leaks=0
```

roberth commented:
Yeah my suggestion to golden-test the generated source was as a poor man's verification. If we can verify memory-correctness automatically, we don't need to golden-test the generated sources.

pennae commented:
> How would we verify that the wrapper didn't modify some unexpected environment variable?

the test can start the wrapper in an empty environment, eg `env -i`

bergkvist committed:
Switch to embedding input arguments instead of generated C code in binary.

tfc commented:
can we also directly aim towards being able to substitute `makeWrapper`? i would really like to be able to have an overlay like `final: prev: { makeWrapper = prev.makeBinaryWrapper; }` and build my systems like that.

roberth commented:
@tfc `makeWrapper` is part of stdenv, so that's a big change that has to go into staging. It's more efficient to complete this work on `master` first and then create a new PR for stdenv targeting `staging`.

bergkvist commented:
In terms of feature parity with makeWrapper (to allow substitution), there are a few things which will not work/be the same:
- `--run "..."`-commands which changes directory, or accesses/changes the environment/sources a bash script
- `--argv0 '$0'` -> Commands which take advantage of how the bash runtime will expand expressions like environment variables.

In general, anything which relies on the inner workings of bash/makeWrapper, rather than strictly on the makeWrapper interface will be hard (and possibly not desirable) to port over.

bergkvist committed:
Separate out indentation responsibility to indent4 in makeCWrapper using awk. 

Generated code no longer needs to worry about its own indent level in the output.

tfc committed:
Add golden effects test

tfc committed:
Fix typo in make-binary-wrapper

blitz commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -863,9 +871,15 @@ someVar=$(stripHash $name)
+
+If you will apply it multiple times, it will overwrite the wrapper file and you will end up with double wrapping, which should be avoided.
+
+### `wrapProgramBinary` \<executable\> \<makeBinaryWrapperArgs\> {#fun-wrapProgramBinary}
```
nit: This sounds like it wraps a program binary, which I find a pretty confusing name. What about `binaryWrapProgram`?

doronbehar replied:
> nit: This sounds like it wraps a program binary, which I find a pretty confusing name. What about `binaryWrapProgram`?

I agree. It's also more consistent with `makeWrapper` vs `makeBinaryWrapper`. I will change it now but this raises the question whether we should make the bash functions have the same name, as @tfc already suggested.

blitz replied:
Having these functions be drop-in replacements for the shell wrappers would be pretty awesome.

tfc replied:
i have just tested building and running some more involved applications with a drop-in like `makeWrapper = makeBinaryWrapper` in an overlay and this seems to work perfectly. (only takes some time because you're rebuilding the world afterward)

doronbehar committed:
wrapProgramBinary -> binaryWrapProgram

Commit:
Inject gcc path into makewrapper script

tfc committed:
make makeWrapper and makeBinaryWrapper drop-in-replaceable

bergkvist committed:
Merge branch 'bergkvist/make-c-wrapper' into make-c-wrapper

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -608,7 +608,9 @@ in
     deps = [ dieHook ];
-  } ../build-support/setup-hooks/make-binary-wrapper.sh;
+  } (runCommand "make-binary-wrapper.sh" {} ''
+      substitute ${../build-support/setup-hooks/make-binary-wrapper.sh} $out --replace " cc " " ${gcc}/bin/cc "
```
I think it'd be safer to use `@cc@ instead of this substitution.

tfc replied:
done

doronbehar committed:
Rephrase documentation for both makeWrapper implementations

tfc commented:
things that we should add:

- `set -euo pipefail` in the script
- `-Werror -Wall -Wpedantic` is a must
- let the user choose/override the compiler
- let the user choose to enable/disable `-fsanitize=address -fsanitize=undefined`

tfc committed:
make-binary-wrapper: Add -Wall -Werror -Wpedantic

tfc committed:
make-binary-wrapper: Add -euo pipefail to bash script

tfc committed:
make-binary-wrapper: Make CC substitution safer

tfc committed:
make-binary-wrapper: Add sanitizer default option

blitz commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -863,9 +865,9 @@ someVar=$(stripHash $name)
 ### `wrapProgram` \<executable\> \<makeWrapperArgs\> {#fun-wrapProgram}
 
-Convenience function for `makeWrapper` that automatically creates a sane wrapper file. It takes all the same arguments as `makeWrapper`, except for `--argv0`.
+Convenience function for `makeWrapper` that replaces `<\executable\>` with a wrapper that executes the original program. It takes all the same arguments as `makeWrapper`, except for `--inherit-argv0` (used by the `makeBinaryWrapper` implementation and `--argv0` (used by both `makeWrapper` and `makeBinaryWrapper` wrapper implementations).
```
```suggestion
Convenience function for `makeWrapper` that replaces `<\executable\>` with a wrapper that executes the original program. It takes all the same arguments as `makeWrapper`, except for `--inherit-argv0` (used by the `makeBinaryWrapper` implementation) and `--argv0` (used by both `makeWrapper` and `makeBinaryWrapper` wrapper implementations).
```

blitz reviewed (approved):
This is really nice. I like it. i wonder whether there is even a performance benefit from using binary wrappers for often-executed wrapped programs.

doronbehar committed:
Small rephrase of wrapProgram documentation

Co-authored-by: Julian Stecklina <js@alien8.de>

pennae commented on pkgs/top-level/all-packages.nix:
```diff
@@ -606,11 +606,20 @@ in
+      makeSetupHook { deps = [ dieHook ]; } script;
+  in
+    lib.makeOverridable f {
+      cc = gcc;
```
```suggestion
      cc = stdenv.cc.cc;
```
defaulting to gcc might cause greatly inflated build closures on darwin

bergkvist replied:
I'm getting this error on my intel macbook pro (2019) after this change:
```sh
$ env NIX_PATH=nixpkgs=$PWD nix-shell -p makeBinaryWrapper --run "makeWrapper /usr/local/bin/python3 a"
bash: /nix/store/5im8ywgl0cilxlysvrwdnhik954nivbm-clang-7.1.0/bin/cc: No such file or directory
```

Seems like `cc` doesn't exist here:
```
$ ls /nix/store/5im8ywgl0cilxlysvrwdnhik954nivbm-clang-7.1.0/bin/
clang			  clang-check	      clang-import-test      clang-rename	   diagtool
clang++			  clang-cl	      clang-include-fixer    clang-reorder-fields  find-all-symbols
clang-7			  clang-cpp	      clang-offload-bundler  clang-tidy		   hmaptool
clang-apply-replacements  clang-format	      clang-query	     clangd		   modularize
clang-change-namespace	  clang-func-mapping  clang-refactor	     cpp		   scan-build
```

pennae replied:
huh, didn't realize clang doesn't come with `cc` links. try `stdenv.cc`, that will *definitely* have a `cc`.

bergkvist replied:
That seems to have worked, but now I get a new error:
```sh
env NIX_PATH=nixpkgs=$PWD nix-shell -p makeBinaryWrapper --run "makeWrapper /usr/local/bin/python3 a"
In file included from <built-in>:362:
<command line>:1:9: error: '_FORTIFY_SOURCE' macro redefined [-Werror,-Wmacro-redefined]
#define _FORTIFY_SOURCE 2
        ^
<built-in>:350:9: note: previous definition is here
#define _FORTIFY_SOURCE 0
        ^
1 error generated.
```

bergkvist replied:
If I remove the sanitizers, the error goes away - so I guess it has something to do with those

bergkvist replied:
Seems to be caused by the address sanitizer

pennae commented on pkgs/top-level/all-packages.nix:
```diff
@@ -606,11 +606,20 @@ in
+  in
+    lib.makeOverridable f {
+      cc = gcc;
+      sanitizers = [ "undefined" "address" ];
```
don't think these should be enabled by default for all packages, only in its own tests. but having the option is very useful.

bergkvist replied:
What are the downsides to having them enabled by default? Will it significantly impact performance/binary size?

tfc replied:
Please leave them enabled by default. I prefer having sanitizers explode loudly instead of having the program doing wild things in case we have undefined behavior triggered in a raw c program. Users who really want to squeeze the last bit performance can still consciously disable this.

pennae replied:
@tfc we agree with you, but sanitizers are a huge overhead for short binaries. we've run a test of a binary that's just ten putenvs, and sanitizers slowed it down 25-fold (0.4ms per run without sanitizers, 10ms with sanitizers enabled). compared to bash wrappers it's still a 3x slowdown (3.5ms for bash). we would argue that that's just *too much* overhead for something that can be exhaustively tested, like the current wrappers can be. if there was an option to include custom code we'd *totally* want those sanitizers to be enabled at all times when custom code is involved though.

andrew-d replied:
FWIW, on my M1 Pro, benchmarked with `hyperfine`:

* No sanitizers: 0.4 ms ± 0.3 ms
* With `-fsanitize=address` alone: 50.6 ms ± 0.8 ms
* With `-fsanitize=undefined` alone: 2.6 ms ± 0.4 ms
* With both: 51.7 ms ± 1.3 ms

So perhaps we can keep `-fsanitize=undefined` by default?

pennae replied:
if it's cheaper than bash it's a win :+1: would still prefer them off by default though. the code is small and readily audited/tested, and we currently trust bash to not attend any shenanigans on odd inputs too.

blitz replied:
Yes, given that these programs do not process input of any kind, I also don't see why the sanitizers should be on by default. Having them in the tests seem sufficient. (This is also in line with how other programs use sanitizers.)

ncfavier replied:
This fails because `stdenv.cc.cc` is unwrapped and has no `/bin/cc`

bergkvist committed:
Change default cc from gcc to stdenv.cc.cc to reduce closure size on darwin

Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

doronbehar commented:
@roberth did you notice we've had 2 review discussions left open? They are an indicator that `makeWrapper` doesn't work on Darwin because of the sanitizers enabled and even on Linux they are a too big performance price to pay for safety.

roberth commented:
@doronbehar I didn't notice the discussion about the failure on macOS, which I should have. The other discussion seemed resolved because this addition is not a performance _regression_ and the trade-off can be adjusted later. Then also the PR already had two approvals and the CI was all-green.

doronbehar commented:
> The other discussion seemed resolved because this addition is not a performance _regression_ and the trade-off can be adjusted later.

It's just that the sanitizing seems to break the build on Darwin as well, besides the `stdenv.cc` issue, as mentioned in @bergkvist comment: https://github.com/NixOS/nixpkgs/pull/124556#discussion_r766237818

What do you think @tfc ?

tfc commented:
> What do you think @tfc ?

So from my perspective there are multiple use cases for this wrapper:

- have *working* wrappers on macos
- have static unscripted wrappers on linux
- have wrappers that are *faster* than bash based wrappers

having sanitizers in the binaries by default provides *working* wrappers (apart from the fact that we need to figure out that one problem on macos with the sanitizers) with a secure default. i think having sanitizers in the tests is not enough because we might have a rest chance to end up with some kind of strange makewrapper flags that lead to an unsafe binary that is not covered by the tests. it's a risk to that then have do strange things instead of loudly exploding.

for the case of "i want fast wrappers" you can simply build a system with the secure ones, see that it works and then rebuild with insecure ones. disabling this security option should most likely be a conscious choice.

bergkvist commented:
Regarding CI, it seems like everything works:
![image](https://user-images.githubusercontent.com/410028/145588708-efaca4d0-cf89-438e-bc83-ab4b80bdca27.png)

But looking more closely at it, it turns out all the tests were skipped:
![image](https://user-images.githubusercontent.com/410028/145588874-50ec633c-242d-4c74-a1fb-0ff6bab13b86.png)

So this also needs to be fixed

bergkvist commented:
I've created a new PR to fix the Darwin issues we discussed now: https://github.com/NixOS/nixpkgs/pull/150079

ncfavier commented on pkgs/top-level/all-packages.nix:
```diff
@@ -606,6 +606,21 @@ in
+      makeSetupHook { deps = [ dieHook ]; } script;
+  in
+    lib.makeOverridable f {
+      cc = stdenv.cc.cc;
```
This fails because `stdenv.cc.cc` is unwrapped and has no `/bin/cc`

ncfavier replied:
I guess this is addressed by #150079, but note that it is also a problem on Linux

roberth commented:
> for the case of "i want fast wrappers" you can simply build a system with the secure ones, see that it works and then rebuild without sanitizers. disabling this security option should most likely be a conscious choice.

Alternatively, you could prove correctness using a _sound_ static analyzer. Also, perhaps the functions could form a memory-safe interface, such that the generated code is as simple as possible. That would make it feasible to have a high assurance without performance impact; without even having to run the analyzer as a build dependency in practice (because the functions have already been checked and that check has been cached).

bergkvist commented:
Yeah, static analyzers are an attractive alternative to using sanitizers - since they move the checking from run-time to build time. This means problems are also detected earlier.

A static analyzer can never prove safety in the general case, since this is an [undecidable/uncomputable problem](https://dl.acm.org/doi/abs/10.1145/161494.161501). So I guess sanitizers might provide a somewhat stronger protection against undefined behaviour (although at a high cost to performance).

It seems like it is common to combine multiple static analyzers - since they tend to specialize in different areas/detect different kinds of problems. When I tested infer static analyzer and the clang static analyzer on the binary wrapper code - there was for the most part no overlap in the problems they would detect.

infer seemed a lot better than clang at detecting buffer overruns and potential NULL dereferencing for example:
 - https://fbinfer.com/docs/all-issue-types#buffer_overrun_l1
 - https://fbinfer.com/docs/all-issue-types#null_dereference

roberth commented:
> A static analyzer can never prove safety in the general case

I don't think we need the general case.

> sanitizers might provide a somewhat stronger protection against undefined behaviour

Only if the analysis is unsound. Undecidability does not exclude soundness and our programs are very simple here, so checking with an analyzer that is sound (but incomplete, as per undecidability) seems feasible.

CodeHawk-C is sound but I don't have experience with it and it hasn't been packaged yet. Maybe other analyzers exist that have the soundness property.

bergkvist commented:
Yes, this is a good point. By having an analyzer that marks everything as potentially unsafe - it would be trivially sound (although not particularly useful). Rather than adding suspicious cases, a sound analyzer would remove things that are provably safe from the pool of potential problems (where everything starts out as potential problems).

bergkvist commented:
It seems like the compilation/tests of makeBinaryWrapper fails on aarch64-darwin, due to some codesign issues. Anyone here with an M1 macbook/that know what this might be about? See the [OfBorg logs](https://logs.nix.ci/?key=nixos/nixpkgs.150079&attempt_id=cbc544bc-16f6-4da2-a853-e1ac962c2387) and  https://github.com/NixOS/nixpkgs/pull/150079 for more info.

therealpxc commented:
Very cool to see this make it into Nixpkgs in some form. Thanks for getting this first version in and for continuing to work to improve this feature! Also thanks to all the reviewers, since this seems like one of the harder changes to review in Nixpkgs and has gotten a lot of work from reviewers. You guys are awesome!

tobiasBora commented:
Since it's time to change the way we handle wrappers, I created a new issue to mention another problem that can be problematic with wrappers (namely that `$0` points in scripts to the wrapping program instead of the original program, which causes troubles if `$0` is used to restart the program later), together with a proposed solution to solve it https://github.com/NixOS/nixpkgs/issues/150841.

Note that as it, this PR does not solve it (nor does https://github.com/NixOS/nixpkgs/pull/150079). Would be great if both issues could be solved at once.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/14

abathur commented:
I haven't had a chance to poke at these wrappers yet, but this seems like a fair place to register that I may need a programmatic way to associate them with their wrapped executables. 

I noted this tentatively back in nixos/rfcs#75, and at the time I was imagining a metadata dotfile alongside the wrapper. I didn't realize there was work under-way here, and the feature I needed this for has been in nixpkgs for a few months. Luckily I noticed toonn mention this on discourse. 😊

Basically:
1. resholve uses [binlore](https://github.com/NixOS/nixpkgs/blob/08b791a01b35f693d8bb57322da498a7e04a6b62/pkgs/development/tools/analysis/binlore/default.nix) to [collect some information about executables](https://github.com/NixOS/nixpkgs/blob/a649cbca0948c39c43c63b790c75b5b6e4db1564/pkgs/development/misc/resholve/resholve-utils.nix#L68).
2. part of this analysis [maps Shell wrappers to wrapped executables by parsing the exec out of the Shell wrapper](https://github.com/abathur/binlore/blob/833d6b1a8731dcf7af40073dd7a5671f5da5b7fa/execers.yall#L14-L41) in order to associate information derived from the executable with the wrapper as well.

ncfavier commented on pkgs/test/make-binary-wrapper/default.nix:
```diff
@@ -0,0 +1,69 @@
+  makeGoldenTest = { name, filename }: stdenv.mkDerivation {
+    name = name;
+    dontUnpack = true;
+    buildInputs = [ makeBinaryWrapper ];
```
I'm adding a cross-compilation test in https://github.com/NixOS/nixpkgs/pull/175317
